File: core-sdk/python/adc_core_sdk/protocol.py
"""Device-plane wire protocol types for ADC (JSON-RPC 2.0 style, MCP semantics).

This module mirrors core-sdk/protocol/protocol.go field by field: JSON field
names, required/optional semantics and defaults must stay identical to the Go
implementation so both languages speak the same wire contract. Contract
fidelity is pinned by the hardcoded sample payloads in tests/test_protocol.py.
"""

from __future__ import annotations

import json
from typing import Any

from pydantic import BaseModel


class WireModel(BaseModel):
    """Base class reproducing Go encoding/json omitempty behaviour on dump.

    Go omits nil pointers, nil maps/interfaces and empty omitempty strings.
    Python models mirror this by excluding None values and any optional string
    fields listed in ``_omitempty_str_fields`` when their value is "".
    """

    #: Optional string fields that Go marshals with omitempty.
    _omitempty_str_fields: tuple[str, ...] = ()

    def to_wire_dict(self) -> dict[str, Any]:
        """Return the wire-compatible dict with Go omitempty semantics applied."""
        data: dict[str, Any] = {}
        for field_name in self.__class__.model_fields:
            value: Any = getattr(self, field_name)
            if isinstance(value, WireModel):
                value = value.to_wire_dict()
            elif isinstance(value, list):
                value = [
                    item.to_wire_dict() if isinstance(item, WireModel) else item
                    for item in value
                ]
            if value is None:
                continue
            if field_name in self._omitempty_str_fields and value == "":
                continue
            data[field_name] = value
        return data

    def to_wire_json(self) -> str:
        """Return compact wire JSON (same shape as Go json.Marshal output)."""
        return json.dumps(self.to_wire_dict(), separators=(",", ":"), ensure_ascii=False)


# ---------------------------------------------------------------------------
# JSON-RPC 2.0 base envelopes
# ---------------------------------------------------------------------------


class JSONRPCError(WireModel):
    code: int
    message: str
    data: Any | None = None

    def __str__(self) -> str:
        return f"jsonrpc error {self.code}: {self.message}"


class JSONRPCRequest(WireModel):
    jsonrpc: str
    id: str
    method: str
    params: Any | None = None


class JSONRPCResponse(WireModel):
    jsonrpc: str
    id: str
    result: Any | None = None
    error: JSONRPCError | None = None


# Standard JSON-RPC error codes (identical to protocol.go).
ERR_PARSE = -32700
ERR_INVALID_REQUEST = -32600
ERR_METHOD_NOT_FOUND = -32601
ERR_INVALID_PARAMS = -32602
ERR_INTERNAL = -32603

# ---------------------------------------------------------------------------
# MCP tool definitions
# ---------------------------------------------------------------------------

#: Default risk level when a device does not report one: approve before use.
DEFAULT_RISK_LEVEL = 2


class MCPTool(WireModel):
    name: str
    description: str | None = None
    inputSchema: Any
    # RiskLevel is the device-reported suggestion (0-3); the cloud DB is
    # authoritative and this value is advisory only (SEC-09).
    riskLevel: int | None = None
    # SchemaVersion is used by the compatibility matrix (SEC-22).
    schemaVersion: str | None = None

    _omitempty_str_fields = ("description", "schemaVersion")

    @property
    def effective_risk(self) -> int:
        """Return the effective risk level, defaulting to 2 when absent."""
        return self.riskLevel if self.riskLevel is not None else DEFAULT_RISK_LEVEL


class ToolsListResult(WireModel):
    tools: list[MCPTool]


class ToolCallParams(WireModel):
    name: str
    arguments: dict[str, Any] | None = None


class ToolContent(WireModel):
    type: str
    text: str | None = None

    _omitempty_str_fields = ("text",)


class ToolCallResult(WireModel):
    content: list[ToolContent]
    isError: bool


# ---------------------------------------------------------------------------
# Method name constants (identical to protocol.go)
# ---------------------------------------------------------------------------

METHOD_TOOLS_LIST = "tools/list"
METHOD_TOOLS_CALL = "tools/call"

File: core-sdk/protocol/protocol.go
// Package protocol 定义 ADC 设备面 wire 协议（JSON-RPC 2.0 风格，MCP 语义）。
// 属于 core-sdk 仓库（Apache-2.0），SDK 与云端共用，禁止引入任何外部依赖。
package protocol

import (
	"encoding/json"
	"fmt"
)

// JSON-RPC 2.0 基础报文

type JSONRPCRequest struct {
	JSONRPC string      `json:"jsonrpc"`
	ID      string      `json:"id"`
	Method  string      `json:"method"`
	Params  interface{} `json:"params,omitempty"`
}

type JSONRPCResponse struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      string          `json:"id"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *JSONRPCError   `json:"error,omitempty"`
}

type JSONRPCError struct {
	Code    int         `json:"code"`
	Message string      `json:"message"`
	Data    interface{} `json:"data,omitempty"`
}

// Error 实现 error 接口，便于网关层统一包装。
func (e *JSONRPCError) Error() string {
	return fmt.Sprintf("jsonrpc error %d: %s", e.Code, e.Message)
}

// 标准 JSON-RPC 错误码
const (
	ErrParse      = -32700
	ErrInvalidReq = -32600
	ErrMethod     = -32601
	ErrInvalidArg = -32602
	ErrInternal   = -32603
)

// MCP 工具定义

type MCPTool struct {
	Name        string          `json:"name"`
	Description string          `json:"description,omitempty"`
	InputSchema json.RawMessage `json:"inputSchema"`
	// RiskLevel 设备上报的建议风险等级（0-3）；云端 DB 为权威，仅作建议（SEC-09）。
	RiskLevel *int `json:"riskLevel,omitempty"`
	// SchemaVersion 工具 schema 版本，用于兼容矩阵（SEC-22）。
	SchemaVersion string `json:"schemaVersion,omitempty"`
}

func (t MCPTool) EffectiveRisk() int {
	if t.RiskLevel != nil {
		return *t.RiskLevel
	}
	return 2 // 缺省先审后用
}

type ToolsListResult struct {
	Tools []MCPTool `json:"tools"`
}

type ToolCallParams struct {
	Name      string                 `json:"name"`
	Arguments map[string]interface{} `json:"arguments,omitempty"`
}

type ToolCallResult struct {
	Content []ToolContent `json:"content"`
	IsError bool          `json:"isError"`
}

type ToolContent struct {
	Type string `json:"type"`
	Text string `json:"text,omitempty"`
}

// 方法名常量
const (
	MethodToolsList = "tools/list"
	MethodToolsCall = "tools/call"
)

// 设备面鉴权握手常量（SEC-03）
const (
	// HeaderXDeviceID 设备码头
	HeaderXDeviceID = "X-Device-ID"
	// HeaderXDeviceTimestamp 认证时间戳（Unix 秒）
	HeaderXDeviceTimestamp = "X-Device-Timestamp"
	// HeaderXDeviceNonce 一次性 nonce
	HeaderXDeviceNonce = "X-Device-Nonce"
	// HeaderXDeviceSignature HMAC(secret, deviceID + timestamp + nonce)
	HeaderXDeviceSignature = "X-Device-Signature"
	// AuthTimeWindowSec 时间戳漂移窗口
	AuthTimeWindowSec = 300
)
